base/namedtuple.jl

```diff
@@ -179,10 +179,11 @@ nextind(@nospecialize(t::NamedTuple), i::Integer) = Int(i)+1
 
 convert(::Type{NT}, nt::NT) where {names, NT<:NamedTuple{names}} = nt
 convert(::Type{NT}, nt::NT) where {names, T<:Tuple, NT<:NamedTuple{names,T}} = nt
-convert(::Type{NT}, t::Tuple) where {NT<:NamedTuple} = NT(t)
+convert(::Type{NT}, t::Tuple) where {NT<:NamedTuple} = (@inline NT(t))::NT
 
 function convert(::Type{NamedTuple{names,T}}, nt::NamedTuple{names}) where {names,T<:Tuple}
-    NamedTuple{names,T}(T(nt))::NamedTuple{names,T}
+    NT = NamedTuple{names,T}
+    (@inline NT(nt))::NT
```

Comment:
Lost T(NT) here?

Reply:
Intentionally - the definition `Tuple(nt) = (nt...,)` unfortunately is not inlinable by our compiler for a mildly-abstract NamedTuple type like we specialize on here (it gets stuck as an `_apply_iterate`)

Also the compiler has a [hardcoded de-optimization](https://github.com/JuliaLang/julia/blob/4cdd864e535b16e928c2eff43e6f1583bd77209d/base/compiler/abstractinterpretation.jl#L2405-L2407) for any calls to `Tuple(...)` - I'm not sure where it comes from but it is a problem for the MWE in the OP

Reply:
Oh, huh, seems like a strange de-optimization there. Is this equivalent? In theory this change is potentially also more semantically correct, since there are many cases where the Tuple T contains a Type, which means the convert to T does not actually return something of type T

Reply:
Seems like @JeffBezanson introduced this problem in
 https://github.com/JuliaLang/julia/pull/33779/files#diff-0fcc5b3f1870b66e0adbebb39ee9ad51f60194ee100f5c51e368c572ade9ca34R800-R801. Maybe we need to revisit that somewhat?

Reply:
Hmm, yes that does seem problematic I agree - probably worth fixing up soon